bot.js

```diff
@@ -172,19 +178,23 @@ function formatDifferenceMessage(difference){
 
 function balance(pool, message){
 
-		/**
-		TODO:
-		Reintroduce Bishop logic
-		Allow leaders to be promoted/demoted and rebalance groups
-		**/
-
-		let [joining, weakest, total] = [pool.slice().sort((a,b) => a.rank-b.rank), [{rank: 100}], [...leaders, ...pool].length]
+		let [bishops, joining, weakest, total] = [pool.filter(member => member.role === "bishop").sort((a,b) => a.rank-b.rank), pool.slice().sort((a,b) => a.rank-b.rank), [{rank: 100}], [...leaders, ...pool].length]
 		leaders.forEach((leader, index) => total > (10*index) ? groups[index] = [leader] : joining.push(leader))
 		while(joining.length){
-			groups.forEach(group => {
+			groups.filter(group => !group.full).forEach(group => {
 				if(totalRank(group) < totalRank(weakest)) weakest = group
 			})
-			weakest.push(joining[joining.length-1]) && joining.pop()
+		if(weakest.length === 10){
+			weakest.full = true
+			weakest = groups.filter(group => !group.full)[0]
+			continue;
+		}
+		weakest.filter(member => member.role === "bishop").length === 0 &&
+		bishops.length
+		  ? weakest.push(bishops[bishops.length - 1]) &&
+		    joining.splice(joining.indexOf(bishops[bishops.length - 1]), 1) &&
+		    bishops.pop()
+		  : weakest.push(joining[joining.length - 1]) && joining.pop();
```

Comment:
`weakest.push(joining[joining.length - 1]) && joining.pop();`
to
`weakest.push(joining.pop())`